<style>
    #faq {
    position: relative;
    }
</style>

<div id='faq'>
    <p><strong>How are features detected?</strong> The submitted plasmid sequence is queried against several databases using <a href="https://www.ncbi.nlm.nih.gov/books/NBK279690/">BLAST</a>, <a href="https://diamond.readthedocs.io">DIAMOND</a>, and <a href="http://eddylab.org/infernal/">Infernal</a>. Features are predicted if there is a match with ≥95% sequence identity to the database. Swiss-Prot hits have a more tolerant minimum sequence identity requiring a ≥50% match and a penalty is applied corresponding to their respective <a href="https://www.uniprot.org/help/annotation_score">annotation score</a>. Matches with significant overlap are filtered so that only the best consensus set of overall features is shown.</p>
    <p><strong>What databases are used?</strong> pLannotate finds nucleotide matches to a set of features provided by <a href="https://www.snapgene.com/">SnapGene</a> and identifies translated nucleotide matches to entries in <a href="https://www.fpbase.org">fpbase</a> and the <a href="https://www.uniprot.org/statistics/Swiss-Prot">Swiss-Prot database</a>. <a href="https://rfam.xfam.org/">Rfam</a> is used to detect non-coding RNAs.</p>
    <p><strong>What is the &quot;%&quot; assigned to annotations?</strong> This is the percentage of the database feature that is matched in the plasmid. It is calculated as the percent match length multiplied by the percent identity within the matching region.</p>
    <p><strong>What do unfilled features on the plot mean?</strong> These are <em>incomplete features</em>; the sequence match in the plasmid covers less than 95% of the full length of the feature in the database.</p>
    <p><strong>Why should I care about incomplete features?</strong> These elements may be leftover fragments from earlier cloning steps used to create a plasmid. If they include only a small fraction of the feature, they likely do not still have the annotated function. However, even small feature fragments may affect plasmid function if they result in cryptic gene expression or are inadvertently combined with other elements during later cloning steps.</p>
    <p><strong>What is &quot;Linear plasmid annotation&quot;?</strong> pLannotate can also annotate engineered DNA sequences that are not circular, such as <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2847241/">pJAZZ</a> linear plasmids, or <a href="https://en.wikipedia.org/wiki/Sanger_sequencing">Sanger sequencing</a> fragments. pLannotate still displays the plot as a circular plasmid map, though the break between the ends is denoted with a heavy black line.</p>
    <p><strong>What is &quot;Detailed plasmid annotation&quot;?</strong> pLannotate can also use a modified algorithm that allows for plasmid features of different types to "nest" within each other. This will likely find even more plasmid features, however, this approach is prone to more false positves potentially requiring more downstream feature curation.</p>
    <p><strong>I encounted a bug, what do I do?</strong> Please contact the author via one of the links below.</p>
</div>
<br>