# prstools


<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->

`prstools` is software to create Polygenic Risk Scores (PRS) directly
from the commandline <br> (and optionally from inside python).

It contains various tools to make PRS generation easier. Installation
and running the demo example should not take more than 10 minutes.

We are actively developing `prstools` and feedback by mail or our
[feedback form](https://forms.gle/TnvNyBX6qDy7Vupn9) (with üèÜ lottery)
is much appreciated!

## Install

To install use the following command.

``` sh
pip install -U prstools
```

It contains a special read-only access key, since this repo is still
private, which can be shared with people that should have access. <br>
If this did not work there is a chance you need to install `pip`, which
can be done most easily with conda by typing `conda install pip`. For
other install issues please check the (install guide is not yet
available) [install guide]() or send us a mail.

## How to use

Immediately after installing `prstools`, it should be possible to
download & run the demo example (~4mb), by pasting the following into
the commandline:

``` bash
# Makes 'example' dir with data in current path:
prstools downloadutil --pattern example --destdir ./; cd example

# Run the model with example data:
prstools prscs2 --ref_dir ldref_1kg_pop --bim_prefix target \
                --sst_file sumstats.tsv --n_gwas 2565 --out_dir ./result 
# prst prscs2 -r ldref_1kg_pop -t target -s sumstats.tsv -n 2565 -o ./result # <- short-form

# Make predictions from weights (plink must be installed):
plink --bfile target --out prspred --score ./result_* 2 4 6 
```

This will run PRS-CS2 on the example data, using the new implementation
to demonstrate the capabilities of `prstools` and shows how you can use
output of `prstools` with plink to make predictions.

<br> More generally you can use `prstools` by typing into the
command-line interface:

``` sh
prstools
```


    Usage:
     prstools <command> ...

    Convenient and powerfull Polygenic Risk Score creation. 
    'prst' is a commandline shorthand for 'prstools'

    Models & Utility Commands:
     <command>
      downloadutil  Download and unpack LD reference panels and other data.
      combine       Yap yap yap new stuff
      xprs          XPRS: A very fast `eXPReSs` infitessimal effect-size
                    prediction method that can estimate heritabilty on the fly.
      prscs2        PRS-CS v2: A polygenic prediction method that infers posterior
                    SNP effect sizes under continuous shrinkage (CS) priors.

<br> By combining `prstools` with another `<command>` a specific model
or other functionality can be used.

``` sh
# The following means exactly the same as 'prstools xprs'
prst prscs2
```


    Usage:
     prst prscs2 [-h  --cpus <number-of-cpus>] --ref <dir/refcode> --target
                       <bim-prefix> --sst <file> --out <dir+prefix>
                       [--n_gwas <num>  --chrom <chroms>]
                       [--colmap <alternative_colnames>  --pred]
                       [--n_iter <n_iter>  --n_burnin <n_burnin>]
                       [--n_slice <n_slice>  --seed <seed>  --a <a>  --b <b>]
                       [--phi <phi>  --clip <clip>  --sampler <sampler>]

    PRS-CS v2: A polygenic prediction method that infers posterior SNP effect sizes under continuous shrinkage (CS) priors.

    General Options:
     -h, --help                               Show this help message and exit.
     -c, --cpus <number-of-cpus>              The number of cpus to use. Generally
                                              most efficient if chosen to be
                                              between 1 and 5. Functionality can
                                              be turned-off completely by setting
                                              it to -1. (default: 1)

    Data Arguments (first 5 required):
     -r, --ref_dir, --ref <dir/refcode>       Path to the directory that contains
                                              the LD reference panel. You can
                                              download this reference data
                                              manually with 'prstools
                                              downloadref'. Soon it will be
                                              possible to automatically download
                                              it on the fly. (required)
     -t, --bim_prefix, --target <bim-prefix>  Specify the directory and prefix of
                                              the bim file for the target dataset.
                                              (required)
     -s, --sst_file, --sst <file>             The summary statistics file from
                                              which the model will be created. The
                                              file should contain columns: SNP,
                                              A1, A2, BETA or OR, P or SE
                                              information. At the moment, the file
                                              is assumed to be tab-seperated, if
                                              you like other formats please let
                                              devs know.Alternative column names
                                              can be specified with --columns
                                              (more info below). SNP column should
                                              contain rsid's. See https://tin
                                              yurl.com/sstxampl for an
                                              example. (required)
     -o, --out_dir, --out <dir+prefix>        Output prefix for the results
                                              (variant weights). This should be a
                                              combination of the desired output
                                              dir and file prefix. (required)
     -n, --n_gwas <num>                       Sample size of the GWAS. Not
                                              required if sumstat has a 'N'
                                              column. (default: None)
     --chrom <chroms>                         Optional: Select specific chromosome
                                              to work with. You can specify a
                                              specific chromosome as e.g. "--chrom
                                              3". All chromosomes are used by
                                              default. (default: all)
     --colmap <alternative_colnames>          Optional: Allows one to specify an
                                              alterative column name for the
                                              internally used columns
                                              snp,A1,A2,beta,or,pval,se_beta,n_eff
                                              (in that order). Forinstance "--
                                              colmap rsid,a1,a2,beta_gwas,,pvalue,
                                              beta_standard_error," (OR & N are
                                              excluded in this example). When the
                                              command is run a quick this_column
                                              -> that_column conversion table will
                                              be shown. Additionaly prstools has
                                              many internal checks to make sure a
                                              good PRS will be generated! By
                                              default this is the PRS-CS standard.
                                              (default: SNP,A1,A2,BETA,OR,P,SE,N)
     --pred, -p                               Optional: Add this argument to
                                              predict the PRS for the induviduals
                                              in the target dataset.

    Model Arguments (all optional):
     --n_iter <n_iter>                        Total number of MCMC iterations.
                                              (default: 1000)
     --n_burnin <n_burnin>                    Number of burn-in iterations if
                                              larger than 1 or fraction of n_iter
                                              if smaller then 1. (default: 0.5)
     --n_slice <n_slice>                      Thinning of the Markov chain.
                                              (default: 1)
     --seed <seed>                            Random seed for reproducibility.
                                              (default: -1)
     --a <a>                                  Parameter a in the gamma-gamma
                                              prior. (default: 1.0)
     --b <b>                                  Parameter b in the gamma-gamma
                                              prior. (default: 0.5)
     --phi <phi>                              Global shrinkage parameter phi. If
                                              phi is not specified, it will be
                                              learnt from the data using a
                                              Bayesian approach (default: -1.0)
     --clip <clip>                            Clip parameter. The default works
                                              best in pretty much all cases.
                                              (default: 1.0)
     --sampler <sampler>                      Sampler algorithm. Rue sampling is
                                              the original sampler, which gives
                                              good results. (default: Rue)

    Examples --> can be directly copy-pasted (:
     prst downloadutil --pattern example --destdir ./; cd example                                                # Makes 'example' dir in current path.
     prstools prscs2 --ref ldref_1kg_pop --target target --sst sumstats.tsv --n_gwas 2565 --out ./result-prscs2 # Run the model with example data.
     prst prscs2 -r ldref_1kg_pop -t target -s sumstats.tsv -n 2565 -o ./result-prscs2 --pred                  # A shorter version of previous that also does the predictions.

As can be seen, there are examples at the end of the help output to
illustrate usage, which should work with a simple copy-paste.

## Contact

For questions and support please send a mail
(menno.j.witteveen@gmail.com).
